proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+::
+
+	"concatMap"   forall next f.   concatMap (\x -> Stream next (f x)) = concatMap' next f
+
+In GHC today, this rule only matches if the target contains a literal application of some function ``f`` to the local variable ``x``.
+This proposal would allow matching the above rule to more complicated targets like ``concatMap (\x. Stream next (x * 2 + x))`` producing ``concatMap' next (\x -> x * 2 + x)``.
+
+*By using more powerful matching, we solve the long-standing problem of fusing
+concatMap under stream fusion.*  In turn, this could
+potentially make stream fusion general enough to replace foldr/build
+fusion in base.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal only changes the semantics of rewrite rules. No new syntax is introduced.
+
+Let us start with some terminology of rewrite rules.
+Consider the rule:
+::
+
+	{-# RULES "wombat"  forall f x.  foo x (\y. f y) = bar x f  #-}
+
+* 	**Template**.
+	The LHS of a rule is called its *template*.
+* 	**Template variables**.
+	The ``forall``'d variables are called the *template variables*.
+	In rule "wombat", ``f`` and ``x`` are template variables.
+* 	**Local binders**.
+	The *local binders* of a rule are the variables bound inside the template.
+	Example: ``y`` is a local binder of rule "wombat".
+	A local binder is specifically not a template variable, nor is it free in the entire rule.
+* 	**Target**.
+	The rule matcher matches the LHS of the rule (the template) against an expression in the program (the *target*).
+* 	**Substitution**.
+	A sucessful match finds a *substitution* S: a binding for each template variable, such that applying S to the LHS yields the target.
+* 	After a successful match we replace the target expression with the substitution S applied to the RHS of the rule.
+
+
+In GHC today, a template variable ``v`` matches any expression ``e`` if
+
+* ``e`` has the same type as ``v``
+* No local binder of the template is free in ``e``.
+
+The change proposed here is that a **higher order pattern** matches any expression (of the same type):
+
+* 	**Higher order pattern (HOP)**.
+	A *higher order pattern* is an expression of form ``f x y z`` where:
+
+	- ``f`` is a *template variable*
+	- ``x``, ``y``, ``z`` are *local binders* (like ``y`` in rule "wombat" above; see definitions).
+	- The arguments ``x``, ``y``, ``z`` are *distinct* variables
+	- ``x``, ``y``, ``z`` must be term variables (not type applications).
+
+* 	A higher order pattern ``f x y z`` matches *any expression* ``e`` provided:
+
+	- The target has the same type as the template
+	- No local binder is free in ``e``, other than ``x``, ``y``, ``z``.
+
+*	If these two condition hold, the higher order pattern ``f x y z`` matches the target expression ``e``, yielding the substitution ``[f :-> \x y z. e]``.
```

Comment:
Doesn't make much sense to have an additional bullet here, in particular because it refers to a sublist in the previous bullet

```suggestion
	If these two condition hold, the higher order pattern ``f x y z`` matches the target expression ``e``, yielding the substitution ``[f :-> \x y z. e]``.
```